### PR TITLE
fix(#36855/#30300): export 404.html correctly

### LINKED ISSUE #30300: 404 page is exported to 404/index.html when trailingSlash is enabled
### What version of Next.js are you using?

11.1.2

### What version of Node.js are you using?

16.12

### What browser are you using?

Firefox

### What operating system are you using?

Ubuntu

### How are you deploying your application?

next export

### Describe the Bug

I don't remember the previous version, but it was working well before. When I updated the `next` version to **11.1.2**, `/about` exports as `/about.html`, that's why the doesn't work.
Then I enabled **trailingSlash**, now the mentioned above case is solved but `404.tsx` is also exported as `404/index.html` which is not expected according to the static servers.

### Expected Behavior

`404.tsx` should be exported as `/404.html` even when **trailingSlash** is enabled.

### To Reproduce

N/A

### PR DESCRIPTION
**Pull request by SukkaW**
## Bug

- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

The PR fixes #30300 and #36855.

The corresponding integration test case has been added.

### CODE REVIEW

File: test/integration/export-404-html/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+.next-dev
```
Comment (huozhi) [inline test/integration/export-404-html/.gitignore:1]: seems we didn't need this ignore in this case?

File: test/integration/export-404-html/test/index.test.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/* eslint-env jest */
```
Comment (huozhi) [inline test/integration/export-404-html/test/index.test.js:1]: do you think we can merge this with existing `test/integration/export`?

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| buildDuration | 15.5s | 15.4s | -35ms |
| buildDurationCached | 6.3s | 6.3s | ⚠️ +12ms |
| nodeModulesSize | 475 MB | 475 MB | ⚠️ +663 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.778 | 3.774 | 0 |
| / avg req/sec | 661.69 | 662.39 | +0.7 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.172 | 1.177 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 2132.7 | 2123.58 | ⚠️ -9.12 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.8 kB | 28.8 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.5 kB | 72.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.7 kB | 2.7 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.71 kB | 5.71 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.65 kB | 2.65 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.9 kB | 15.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| buildDuration | 17.5s | 17.5s | -34ms |
| buildDurationCached | 6.3s | 6.2s | -43ms |
| nodeModulesSize | 475 MB | 475 MB | ⚠️ +663 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.81 | 3.774 | -0.04 |
| / avg req/sec | 656.15 | 662.38 | +6.23 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.157 | 1.17 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 2161.62 | 2137.11 | ⚠️ -24.51 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.3 kB | 29.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 73.2 kB | 73.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.72 kB | 2.72 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.8 kB | 5.8 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.77 kB | 2.77 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js fix-36855 | Change |
| - | - | - | - |
| index.html gzip | 527 B | 527 B | ✓ |
| link.html gzip | 542 B | 542 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.59 kB | 1.59 kB | ✓ |

</details>

</details>
Commit: e5f909aaffc7ad7d5ca6eec97d95ce1dbab50473

File: test/integration/export-404-html/test/index.test.js (hunk 2 above)
Comment (SukkaW) [reply test/integration/export-404-html/test/index.test.js:1]: > do you think we can merge this with the existing `test/integration/export`?

The current `test/integration/export` does export tests in one pass: first build and export a version with `trailingSlash` enabled, then export a version with `trailingSlash` disabled output to the same directory (in `beforeAll`). That's why the test case I added in #36827 passed.

File: test/integration/export-404-html/.gitignore (hunk 1 above)
Comment (SukkaW) [reply test/integration/export-404-html/.gitignore]: Fixed in [`08214a9` (#36910)](https://github.com/vercel/next.js/pull/36910/commits/08214a97e986d939737119ada7582a221cfb8b79)

File: test/integration/export-404-html/test/index.test.js (hunk 2 above)
Comment (huozhi) [reply test/integration/export-404-html/test/index.test.js:1]: we can leverage the `export-override-404` integration test

- rename it to `export-404`
- 1 case of exsiting overriding 404 case
- another 2 cases with current trailing slash enabled and disabled cases

the purpose is to group the tests together for better readability and maintaining

Comment (SukkaW) [reply test/integration/export-404-html/test/index.test.js]: Done in [`de95c82` (#36910)](https://github.com/vercel/next.js/pull/36910/commits/de95c826e3541ed9a72685e7c514dd03b008f9ab)

**Review by huozhi — APPROVED** [review]
Thank you!

**Pull request merged**